src/persist-client/src/internal/state_versions.rs

```diff
@@ -641,9 +714,24 @@ impl StateVersions {
         T: Timestamp + Lattice + Codec64,
         D: Semigroup + Codec64,
     {
+        let latest_rollup_seqno = *state.latest_rollup().0;
+
+        let mut verify_seqno = latest_rollup_seqno;
+        for diff in &diffs {
+            assert_eq!(verify_seqno.next(), diff.seqno);
```

Comment:
`[latest_rollup_seqno, state_seqno]` or `[latest_rollup_seqno, state_seqno)`?

Reply:
i was taking about `[latest_rollup_seqno` instead of `(latest_rollup_seqno`. i think `state_seqno]` (what you have) is already correct

Reply:
that means we store the diff for a rollup twice, right? once as `state_seqno]` on the rollup itself, and then `[latest_rollup_seqno,` on the next rollup

I don't think that's a problem, but I had been thinking it'd be nice if the diffs across a consecutive series of rollups exactly corresponded to what was in Consensus.

Reply:
yeah, that's fair. and i also think the `[latest_rollup_seqno` would be hard (not possible?) to do for the initial rollup

note that because of the invariant i mentioned above, [we already have to special case ==](https://github.com/MaterializeInc/materialize/blob/6daf1aa3cd67ee3cb2cf5a4674887c2ff09a77b6/src/persist-client/src/internal/state_diff.rs#L345-L347) when applying a series of diffs, so even if there's a bit of duplication in the diffs stored inline in the rollups, you don't have to like trim the dups or anything. blindly applying them all will work

> but I had been thinking it'd be nice if the diffs across a consecutive series of rollups exactly corresponded to what was in Consensus.

so this is in some sense exactly also my goal here, we're just focusing on different parts of it :D! right now if the earliest live diff is X and the current seqno is Y, then we'll have a rollup at X and in consensus we'll have `[X,Y]` (where the X diff is not necessary). if we store `(latest_rollup_seqno` inline in the diffs, we'd end up with `(X, Y]` but would have no duplication if you walk through a series of aligned rollups. like I said, the X diff is not necessary, but my handwave gut feeling is that (given we already have to no-op applying diff X to in-mem state at version X) that it's better to prioritize having the same set of diffs available over not having the internal duplication

all that said, this is very much a judgement/predictive call and i'm very much willing to be convinced. i'd love to hear what @bkirwi thinks here

Reply:
Before now I did not think anything about this! Anyways here's a fresh mental model:
- Imagining the seqno space as a timeline, going from 0 (the minimum seqno) to some number large enough that we'll never have to think about it.
- States are points on that line. (Consequentially, so are rollups.)
- Diffs are intervals on the line... the set of changes between two states. As an interval, each diff has a start seqno and an end seqno... but in many cases we identify them just by the end seqno, such as in the `VersionedData` struct.
  - It seems unambiguous to identify a diff by its endpoint, since consensus linearizes all writes, so the start point is just the endpoint of the previous diff (if there was one) or zero (if there was not). (Also I suspect most diffs have length one, but I don't know if it's a good idea to rely on that)
  - This also makes the natural convention for seqno ranges closed-open instead of the [dijkstrian open-closed convention](https://www.cs.utexas.edu/~EWD/transcriptions/EWD08xx/EWD831.html). I find this slightly harder to think about but it seems to cause no real problem. 

So concretely, if we're writing a rollup at v8 and the most-recent rollup was at v4, we'll need diffs v5, v6, v7, v8. And we're discussing whether we'd also want to keep around the v4 diff.

Anyways:
- Sounds like we all agree on the set of diffs that are necessary to preserve.
- One advantage of having an "extra" diff is that it makes it easier to figure out the actual start of the range. (It's the seqno of the first diff in state!) It may be worth writing the range to the rollup explicitly.
- I'm sympathetic to the idea that rollups should not overlap. I'm also sympathetic to the idea that rollups can overlap arbitrarily, and the important thing is that none of the range is uncovered. I'd be a bit leery of the idea that we want to encode that rollups should overlap by exactly one diff.